src/lib/select/select.ts

```diff
@@ -441,65 +513,77 @@ export class MdSelect implements AfterContentInit, ControlValueAccessor, OnDestr
   /** Sets up a key manager to listen to keyboard events on the overlay panel. */
   private _initKeyManager() {
     this._keyManager = new FocusKeyManager(this.options);
-    this._tabSubscription = this._keyManager.tabOut.subscribe(() => {
-      this.close();
-    });
+    this._tabSubscription = this._keyManager.tabOut.subscribe(() => this.close());
   }
 
   /** Drops current option subscriptions and IDs and resets from scratch. */
   private _resetOptions(): void {
     this._dropSubscriptions();
     this._listenToOptions();
     this._setOptionIds();
+    this._setOptionMultiple();
   }
 
-  /** Listens to selection events on each option. */
+  /** Listens to user-generated selection events on each option. */
   private _listenToOptions(): void {
-    this.options.forEach((option: MdOption) => {
-      const sub = option.onSelect.subscribe((event: MdOptionSelectEvent) => {
-        if (event.isUserInput && this._selected !== option) {
-          this._emitChangeEvent(option);
+    let source = Observable.merge.apply(Observable,
+        this.options.map(option => option.onSelectionChange));
+
+    this._optionSubscription = source
+      .filter((event: MdOptionSelectionChange) => event.isUserInput)
+      .subscribe((event: MdOptionSelectionChange) => {
+        let wasSelected = this._model.isSelected(event.source);
+
+        if (this.multiple) {
+          this._model.toggle(event.source);
+        } else {
+          this._clearSelection(event.source);
+          this._model.select(event.source);
+        }
+
+        if (wasSelected !== this._model.isSelected(event.source)) {
+          this._propagateChanges();
+        }
+
+        this._setValueWidth();
+
+        if (!this.multiple) {
+          this.close();
         }
```

Comment:
This method is getting pretty fat since the `_onSelect()` method was removed. Can you re-add it to move some of this logic out of the main function (lines 535-552)?